implementation-details.md

```diff
@@ -244,14 +258,24 @@ where:
 * their stakes are summed up starting from the highest score node and going down the list to nodes with lower score
+* we ensure at least 2 nodes are always kept: `if > 1 && ...`
 * when the sum of stakes reaches a defined fraction of the node's own stake: `sum > min(node_stake, origin_stake) * 0.15`, we store the current loop iteration index (`pos = i`) and finish the loop
 * all left nodes whose stake was not summed up yet, starting from `pos + 1`, are pruned.
 
+### Pruning nodes
+
+When a node wants to prune some of its peers it sends a prune message. We already know that pruning is based on the receive cache score and peer stake. In the example posted above node G has a low score in the receive cache of node C. Therefore node C sent a prune message to G with a list of pruned nodes containing A, which means - hey G don't send me messages originating from A anymore.
+
+
+The node receiving the prune message will update its active set of nodes and add pruned nodes to the bloom filter of the message origin entry. In the above example, node G receiving the prune message will add A into the bloom filters of node C in the active set of nodes. Then, whenever G would like to send a push message to C will first check which message origins are pruned by C by checking the bloom filter. If the push message comes from A, G will not send it to C.
+
 ### Pull requests
 
-Nodes are constantly sending pull requests to obtain data they don't contain. To avoid receiving data the node already contains, it creates a list of `CrdsFilter`s which are inserted into pull requests. Each `CrdsFilter` contains a bloom filter with hashes of data the node already possesses - `CrdsValue`s from `crds`, purged values, and failed inserts. Bloom filter is a very fast filter used for checking whether it contains particular data. It has however some range of false positives - it can say a given data exists in the filter when it actually isn't.
+Push message propagation doesn't guarantee that all nodes in the network receive every message. As a result, nodes need a way to ask their peers for any information they do not have. A node doesn't know what it doesn't have. So, the node sends the pull request with data it already has to its peers. The peers will then look over their `crds` table and respond with any data the requesting node does not have.
+
+To avoid receiving data the node already contains, it creates a list of `CrdsFilter`s which are inserted into pull requests. These filters are used by peers to quickly find out what data the requesting node is missing. Each `CrdsFilter` contains a bloom filter with hashes of data the node already possesses - `CrdsValue`s from `crds`, purged values, and failed inserts. Bloom filter is a very fast filter used for checking whether it contains particular data. It has however some range of false positives - it can say a given data exists in the filter when it actually isn't.
```

Comment:
"When creating a pull request, a node needs a way to ensure it does not receive back data it already contains. To do this, the node creates a list of ..."

Reply:
> when it actually isn't.

"...when it actually doesn't"